File: backend/app/seeds/rate_tiers.py
from decimal import Decimal

from app.extensions import db
from app.models.shipping_rate_tier import ShippingRateTier
from app.services.shipping_service import usd_for_billable_lbs

RATE_TIER_SEED = [
    (f"{lbs} lb" if lbs == 1 else f"{lbs} lbs", lbs, lbs, "flat", usd_for_billable_lbs(lbs), None)
    for lbs in range(1, 31)
]


def seed_rate_tiers():
    if ShippingRateTier.query.count() > 0:
        return

    _insert_rate_tiers()


def replace_rate_tiers():
    ShippingRateTier.query.delete()
    db.session.commit()
    _insert_rate_tiers()


def _insert_rate_tiers():
    for i, (label, min_w, max_w, ptype, flat, per_lb) in enumerate(RATE_TIER_SEED, start=1):
        tier = ShippingRateTier(
            display_label=label,
            min_weight_lbs=min_w,
            max_weight_lbs=max_w,
            pricing_type=ptype,
            flat_rate_usd=flat,
            rate_per_lb_usd=per_lb,
            sort_order=i,
            is_active=True,
        )
        db.session.add(tier)

    db.session.commit()

File: server/src/seeds/rateTiers.ts
import { count } from 'drizzle-orm'
import { db } from '../db/index.js'
import { shippingRateTiers } from '../db/schema/index.js'
import { MAX_AUTO_RATE_LBS, REVISED_RATE_USD_BY_LBS } from '../data/revisedRateTable.js'

function buildRateTierSeed() {
  return Array.from({ length: MAX_AUTO_RATE_LBS }, (_, i) => {
    const lbs = i + 1
    return {
      label: lbs === 1 ? '1 lb' : `${lbs} lbs`,
      minW: lbs,
      maxW: lbs,
      flat: REVISED_RATE_USD_BY_LBS[lbs]!,
    }
  })
}

const RATE_TIER_SEED = buildRateTierSeed()

async function insertRateTiers() {
  for (let i = 0; i < RATE_TIER_SEED.length; i++) {
    const { label, minW, maxW, flat } = RATE_TIER_SEED[i]!
    await db.insert(shippingRateTiers).values({
      displayLabel: label,
      minWeightLbs: minW,
      maxWeightLbs: maxW,
      pricingType: 'flat',
      flatRateUsd: flat.toFixed(2),
      ratePerLbUsd: undefined,
      sortOrder: i + 1,
      isActive: true,
    })
  }
}

export async function seedRateTiers() {
  const [{ value: tierCount }] = await db.select({ value: count() }).from(shippingRateTiers)
  if (tierCount > 0) return
  await insertRateTiers()
}

export async function replaceRateTiers() {
  await db.delete(shippingRateTiers)
  await insertRateTiers()
}
